crates/testing/src/consistency_task.rs

```diff
@@ -0,0 +1,176 @@
+#![allow(clippy::unwrap_or_default)]
+use std::collections::BTreeMap;
+
+use anyhow::{bail, ensure, Context, Result};
+use async_trait::async_trait;
+use committable::Committable;
+use hotshot_types::{
+    data::Leaf,
+    event::{Event, EventType},
+    traits::node_implementation::NodeType,
+};
+
+use crate::{
+    overall_safety_task::OverallSafetyPropertiesDescription,
+    test_task::{TestResult, TestTaskState},
+};
+
+trait Validatable {
+    fn valid(&self) -> Result<()>;
+}
+
+pub type NodeMap<TYPES> = BTreeMap<<TYPES as NodeType>::Time, Vec<Leaf<TYPES>>>;
+
+pub type NodeMapSanitized<TYPES> = BTreeMap<<TYPES as NodeType>::Time, Leaf<TYPES>>;
+
+pub fn sanitize_node_map<TYPES: NodeType>(
+    node_map: &NodeMap<TYPES>,
+) -> Result<NodeMapSanitized<TYPES>> {
+    let mut result = BTreeMap::new();
+
+    for (view, leaves) in node_map.iter() {
+        let mut reduced = leaves.clone();
+
+        reduced.dedup();
+
+        match reduced.len() {
+            0 => {}
+            1 => {
+                result.insert(*view, reduced[0].clone());
+            }
+            _ => {
+                bail!(
+                    "We have received inconsistent leaves for view {view:?}. Leaves:\n\n{leaves:?}"
+                );
+            }
+        }
+    }
+
+    Ok(result)
+}
+
+/// For a NodeLeafMap, we validate that each leaf extends the preceding leaf.
+impl<TYPES: NodeType> Validatable for NodeMapSanitized<TYPES> {
+    fn valid(&self) -> Result<()> {
+        let leaf_pairs = self.values().zip(self.values().skip(1));
+
+        // Check that the child leaf follows the parent, possibly with a gap.
+        for (parent, child) in leaf_pairs {
+            ensure!(
+              child.justify_qc().view_number >= parent.view_number(),
+              "The node has provided leaf:\n\n{child:?}\n\nbut its quorum certificate points to a view before the most recent leaf:\n\n{parent:?}"
+            );
+
+            if child.justify_qc().view_number == parent.view_number()
+                && child.justify_qc().data.leaf_commit != parent.commit()
+            {
+                bail!("The node has provided leaf:\n\n{child:?}\n\nwhich points to:\n\n{parent:?}\n\nbut the commits do not match.");
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub type NetworkMap<TYPES> = BTreeMap<usize, NodeMap<TYPES>>;
+pub type NetworkMapSanitized<TYPES> = BTreeMap<usize, NodeMapSanitized<TYPES>>;
+
+pub fn sanitize_network_map<TYPES: NodeType>(
+    network_map: &NetworkMap<TYPES>,
+) -> Result<NetworkMapSanitized<TYPES>> {
+    let mut result = BTreeMap::new();
+
+    for (node, node_map) in network_map {
+        result.insert(
+            *node,
+            sanitize_node_map(node_map)
+                .context(format!("Node {node} produced inconsistent leaves."))?,
+        );
+    }
```

Comment:
This could be `BTreeMap::from_iter` if you wanted it to be.

Reply:
I initially tried something like that, but I don't think it quite works because you can't get the short-circuiting/early return behavior from the `?` here without an explicit loop